### DISCUSSION TITLE
tutorials/plugins/editor/making_plugins

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#137] [User-contributed notes]
# tutorials/plugins/editor/making_plugins

About plugins: A plugin is a great way to extend the editor with useful tools. It can be made entirely with GDScript and standard scenes, without even reloading the editor. Unlike modules, you don'...

https://docs.godotengine.org/en/stable/tutorials/plugins/editor/making_plugins.html

**Comment by Rivers47** [comment]
Why does the C# version of MyButton doesn't have the enclosing #if TOOLS #endif with it?

**Reply by paulloz** [reply]
Hi. If that class was wrapped by `#if TOOLS`, the node type would not be compiled when the game is exported. You can find more information about it [here](https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_features.html#full-list-of-defines)

**Comment by Ceruin** [comment]
When I try in C# my plugin.cfg auto-gens to:
```
[plugin]

name="test123"
description=""
author="Fuey500"
version=""
script="test123.cs"
```
which is not what the 4.4 stable tutorial shows.

**Reply by Calinou** [reply]
Are you referring to this part of the tutorial?

<img width="433" height="145" alt="image" src="https://github.com/user-attachments/assets/70e3fa75-a865-458d-b952-6b20efb9626d" />

The values listed here are just meant to show what you should enter in the dialog, not what `plugin.cfg` contains.

**Reply by Ceruin** [reply]
Yes, Ah sorry I thought it was a representation of the file.

**Comment by Moosh64** [comment]
Making a note because it gave me a bit of headache figuring this out and no tutorial I found mentioned it: it says the root node of a dockable dialog should be Control or one of its child classes, but what you really want is the ScrollContainer specifically. This enables scrollbars when elements get bigger than the dock

**Comment by HaruYou27** [comment]
Do NOT use `add_custom_type()` together with `class_name`. This will cause

```
ERROR: Failed parse script foobar.gd
Class FooBar hides a global script class.
```
I downloaded some plugins from the Godot Asset Library and they all have this issues. Perhaps the docs should warns about this ?

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/11507

**Reply by Nodragem** [reply]
I was actually going to ask why there is no `add_custom_type` and `remove_custom_type` in the enter and exit tree functions. That makes things so much easier!

**Comment by hrharkins** [comment]
Since 4.6.dev5 it appears that the interface will not allow creating nodes from add_custom_type if the script does not have a matching class_name: https://github.com/godotengine/godot/issues/114349

**Reply by Calinou** [reply]
This should be fixed by https://github.com/godotengine/godot/pull/114364 now.

**Reply by hrharkins** [reply]
It is indeed.  Thanks!

**Comment by Gonderage** [comment]
When should something be done in the `enable_plugin` and `disable_plugin` versus the `enter_tree` and `exit_tree`? The two method pairs only seem to differ on who can do tree-related operations, in which case the `*_plugin` methods seem redundant.

**Reply by XandaPanda42** [reply]
As far as I'm aware (new to plugin dev), it depends on what the extension does. Like most things in the node structure it's just timing. `enable_plugin` is probably called right after you click "okay" in the plugin menu.  But the order that `enter_tree` happens in is dependent on a lot of other things, like the extension's position in the tree.

You **can** put most of the startup code in `enter_tree`, and ignore `enable_plugin`, and it'll most likely work fine for simpler plugins. For more complex stuff like having several plugins and sub plugins, it's best to use `enable_plugin` for stuff you want to happen BEFORE any of the plugins enter the tree.

Also, if you're making an import plugin, or adding a custom server or something, and the code in `enter_tree` relies on resources existing before then, or has another plugin as a dependency, you might want disable the plugin if those resources aren't available. The code to do that would go in `enable_plugin` because `enter_tree` might be too late.

TDLR (from a not expert): `enter_tree` for scene setup and plugin UI and stuff, `enable_plugin` for anything you need to be guaranteed to be done before then. Loading scripts, making sure other plugins and classes exist, adding or creating resources in the RenderingServer etc.`

**Comment by bleuthoot-sven** [comment]
The documentation for C# mentions to use `AddControlToDock` to add a dock control. However, the IDE notifies that the method is deprecated.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/11724

**Comment by Spike0471** [comment]
The "AddControlToDock" is deprecated, the dock code should be:
```csharp
public override void _EnterTree()
    {
        // Initialization of the plugin goes here.
        var dockScene = GD.Load<PackedScene>(
                "xxxxx.tscn"
            )
            .Instantiate<Control>();
        _dock = new EditorDock();
        _dock.AddChild(dockScene);
        _dock.Title = "Dock";
        _dock.DefaultSlot = EditorDock.DockSlot.Bottom;
        _dock.AvailableLayouts = EditorDock.DockLayout.Horizontal | EditorDock.DockLayout.Floating;
        AddDock(_dock);
    }
```

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/11724

**Comment by thiscris** [comment]
Hey, I want to say that I spent a lot of time trying to get a plugin with a Sub Viewport to work. I even commented on a bug reported in 2023. https://github.com/godotengine/godot/issues/81957 . I was able to solve it, eventually. I think that it would be useful to have something mentioned about it here.

**Reply by XandaPanda42** [reply]
I think the first two parts have been mentioned, but could definitely use some clarity. 

Every script in a plugin needs to be a tool script. If the script isn't marked with `@tool`, the plugin will read it as an empty script.

The mouse filter part is common to all Control type nodes, so I think it's assumed that it is known already. It's an issue with GUI development in Godot in general. 

The event forwarding part I can't make heads or tails of though. I don't get why it works that way yet as I'm still learning the event and GUI system, but that definitely would be helpful to know beforehand. 

Something else that might be good to know for plugin scripts is that you can't actually create a destructor in a RefCounted or Resource script. The NOTIFICATION_PREDELETE doesn't get called for them at all (in the source it just gets skipped) so if you're using the servers or plain objects in your plugin, make sure you're cleaning up everything before you remove the last reference to it. Tool scripts are quite prone to memory issues so it's something to watch out for.

**Reply by thiscris** [reply]
Thanks. That is a valuable insight. There isn't any way to profile their performance, is there?

**Comment by passiflora-brumaria** [comment]
In order for custom Node types registered this way to have a description in the "Create New Node" UI, they will also need class_name. The addon folder itself will serve as a scope so that these types don't pollute the UI from disabled addons, even if they have class_name. That said, custom Resource types with class_name in disabled addons will still show up everywhere in the project (as of 4.5.1.stable).

**Reply by XandaPanda42** [reply]
Just adding some extra info, I'd need to double check but I'm pretty sure if you prefix your custom class_name with `Editor` (eg. EditorTextBox, EditorDisplayPanel, etc), the node type won't show up in the "Add New Node" dialog either, so it's another good way to keep some of the background addon stuff separate, though it only works for the Node menu as far as I've noticed. It's also just good practice to do to help you distinguish addon related classes from game related ones.

Any Resources or PackedScenes in the Addon folder also won't show up by default when selecting `Quick Load` on an exported property, but you can still search the folder by toggling the `Addons` checkbox in the bottom right.

**Comment by BaconEggsRL** [comment]
Are UIDs supposed to be uploaded with plugins? Or should they be ignored? I got a bunch of UID errors trying to install my plugin from the asset lib for the first time. Thanks!
